### ISSUE TITLE
On the road to a 1.0.0 release

### ISSUE THREAD

**Issue by NicolasRitouet** [deployd/deployd#761]
Hi community,

Following the discussion with @houmark on the ticket https://github.com/deployd/deployd/pull/753, we would like to have your opinion on releasing a 1.0.0 version.
Many people are already using deployd on production, but we (the maintainer's team) don't think that it's quite ready for this, notably on the security/documentation aspect.

**We might use this opportunity to make breaking changes (like using mongoDB ids instead of generating our own ids)**

What do you folks think about this?

**Comment by andreialecu** [comment]
Would you care to elaborate on the security aspect? I don't know of any security issues as long as you take care to lock down sensitive event scripts properly (eg. `if (!me || !me.admin) cancel(...);`)

**Comment by andreialecu** [comment]
There are many things missing from a full 1.0 version. We'd need to update the documentation at a minimum.

Also, there are some unfortunate decisions in the way it was initially architected, which I'm working around, but are not very obvious to a new user. Things like:

- Scripts for PUT/POST and validate reading the item from the database and saving the full item every time, resulting in possible race conditions when multiple saves happen on an item cocurrently. An officially supported way to call findAndModify would be required here. It's possible to call findAndModify through https://github.com/deployd/deployd/pull/608 but I'm not quite happy with that.

- Aggregation is possible, but also through that #608 PR

- IDs should probably be native mongo ObjectIDs created by deployd instead. This could be a breaking change though.

- There needs a clear way to define custom library code which can be reused, otherwise the event scripts become spaghetti code. The workaround here is to instead require() code defined in some common './lib/' folder from event scripts. But I think a separate resource type that is just for Code would be a better option.

- The dashboard should be remade with Angular2 or React, it's running on an old version of KnockoutJS making it hard to maintain.

I use deployd because it's easy to bend to my needs, and I know of several other projects using deployd as well. I can work around any of the shortcomings, but a new user might be turned off by them, since none of the workarounds are very obvious or well documented.

**Comment by safaorhan** [comment]
I'm using deployd for production, it's stable enough for my needs. I support the 1.0.0 release 👍

**Comment by NicolasRitouet** [comment]
> Would you care to elaborate on the security aspect? I don't know of any security issues as long as you take care to lock down sensitive event scripts properly (eg. if (!me || !me.admin) cancel(...);)

True, we (you!) fixed many things related to security, but we would need to improve the documentation to avoid having unsecured instances of deployd online.
Thanks for your valuable input.

**Comment by NicolasRitouet** [comment]
> There are many things missing from a full 1.0 version. We'd need to update the documentation at a minimum.

I agree 👍  Should be doable.

> Scripts for PUT/POST and validate reading the item from the database and saving the full item every time, resulting in possible race conditions when multiple saves happen on an item cocurrently. An officially supported way to call findAndModify would be required here. It's possible to call findAndModify through #608 but I'm not quite happy with that.

What would you make you happy with this?

> Aggregation is possible, but also through that #608 PR
> 
> IDs should probably be native mongo ObjectIDs created by deployd instead. This could be a breaking change though.
Since we are changing major version number, breaking changes are more than ok. It should actually be the perfect time to do other necessary breaking changes.

> There needs a clear way to define custom library code which can be reused, otherwise the event scripts become spaghetti code. The workaround here is to instead require() code defined in some common './lib/' folder from event scripts. But I think a separate resource type that is just for Code would be a better option.
Interesting, I don't have this use-case, but I'd support this feature.

> The dashboard should be remade with Angular2 or React, it's running on an old version of KnockoutJS making it hard to maintain.
Also agree, I'll create a ticket to find contributors for this ticket.

> I use deployd because it's easy to bend to my needs, and I know of several other projects using deployd as well. I can work around any of the shortcomings, but a new user might be turned off by them, since none of the workarounds are very obvious or well documented.

Let's start by listing the necessary workaround and see if we need to document them or fix them.

**Comment by andreialecu** [comment]
I think the documentation should be part of this project, rather than being separate. And every submitted PR should be accompanied by documentation changes.

That would make it much easier to be kept in sync, and it's a lot easier to work just on one project rather than two. 

Unfortunately I don't have the time to look up what would be the best way to do this at the moment, so if someone wants to help - it would be amazing.  The easiest solution would be to just store a bunch of `.md` files in some `docs` folder under this project, since they're already easily accessible via GitHub, but any other ideas are welcome.

**Comment by NicolasRitouet** [comment]
> The dpd cli should be deprecated/removed and a deployd app should be ran as a regular node app instead. Also already possible, but not documented properly.

I also agree.
I also think we should deprecate the windows & mac installers.
I was against it a while ago because installing node on windows used to be a pain, now that it works really good (and easily), I would remove these.
Anyone against it?

**Comment by hjanuschka** [comment]
we use deployed from current master in production under heavy load.
i would love to see the following:
  * no breaking changes, or atleast not too much, and good documented
  * shared code per collection (right now we use custom npm modules for shared code)
  * shared code per instance (right now we use custom npm modules for shared code)
  * maybe dpd as express middleware should be the preffered one? atleast we figured out its a bit mor performant, as we could also use a `staticServe` middleware for jpgs and so on.
  * better dashboard
    * filter
    * delete all records
    * export to CSV/jsonArray
    * sorting



overall i love this project, the idea, the implementation, and i appreciate the work from all of you guys, but i hope that a 1.0.0 won't break too much ;)

**Comment by houmark** [comment]
I don't think deployd is truly using semver or respecting it.

From the semver docs:

> If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backwards compatibility, you should probably already be 1.0.0.

**Used in production:** Yes
**Stable API:** Yes
**Should worry about backwards compatibility:** Yes

Deployd has been sub 1.0.0 for several years but it's probably being used or has been used successfully in hundreds if not thousands of projects and while docs are suffering and there's other challenges, the only way to ensure that breaking changes can be done in the future in a correct semver way, is by releasing 1.0.0 and then follow semver truly for features and breaking changes so users of deployd can be warned before doing an update (I would recommend a maintainer commit policy based on [semantic-release](https://github.com/semantic-release/semantic-release) which also takes care of changelogs and outlining breaking changes).

A version 1.0.0 could also help adoption rate and with adaption rate often comes contributions either in form of code or sponsoring of features. If the deployd maintainers pushes documentation as one of the most important "features", then that can also be sponsored.

I like to use [npm-check](https://www.npmjs.com/package/npm-check) for checking for updated packages in my projects and when you do that with deployd, it will simply show that it's Non-Semver and thus I cannot know if this release is breaking or not. The only way to "fix" that is by taking deployd to 1.0.0 and then start working on the nice to have features and respect semver in each release.

In relation to security. There's no known security issues with the deployd architecture. Deployd's base functionality is to build public API's and it has security features to lock down resources by need using `cancel` / `cancelUnless(!me || !internal)` etc. and those are documented to a pretty okay level already I think. It would be nice to have a "non-public" by default global feature, but that can also be added to the long wishlist :)

Please refrain as a maintainer to write stuff like:

> ....don't think that it's quite ready for this, notably on the security/documentation aspect

This only makes potential adopters think deployd is insecure by design and that's simply not true.

I also struggle with race situations when saving, I also would like to see an improved dashboard, true MongoDB based ID's and aggregation but all of those features can be 2.0, 3.0 etc.

**Comment by NicolasRitouet** [comment]
@houmark 
Thanks for your input, after discussing with the team, we decided to release a v1.0.0 soon.
The only requirement is to deprecate the windows/macosx installers, as they might bring too many bug ticket (it's in progress)
The rest of the roadmap will be for the post v1.0.0 version.

**Comment by andreialecu** [comment]
Sub 1.0 doesn't necessarily mean non production approved.

Node has been at 0.x.x for a few years before actually jumping to version 4.

react-native is at version 0.36 and is being used in production for a few years as well.

**Comment by houmark** [comment]
[React](https://github.com/facebook/react/blob/master/CHANGELOG.md) is version 15.x and seems to be respect semver though :)

Since Node.js went +1.x it has jumped to version 7.x in a very short time though, also because they are following the semver concept in full.

Just because others *did* it wrong (and mostly corrected it since) doesn't mean Deployd should keep doing it.

For me it's not about non-production or production "approved" - for some users of deployd there will be missing features that makes it non-production capable maybe. For me it's about maintenance and being able to upgrade safely and with peace in mind because breaking changes will be easier to spot and won't slip in, in a minor release while minor patch fixes can be updated fully automatic by ~ in package.json. When the version is below 1.x and crawling slowly it's pretty much impossible to respect semver as a major breaking change requires going from v1.x to v2.x. 

There's a lot of feature wishes on the table and some of those may be in some way breaking, but following semver will actually make those easier to handle IRL, because if people have the correct configuration in their package.json (and if they don't they are asking for trouble), then they won't be automatically upgrading to a breaking change and they will be forced to manually review the changelog and fully test and adjust their code if needed, before they upgrade.

As Stephan Bönnemann (who is arguing hard for semver and did talks on the subject) said in one of the videos I posted (loosely quoted):

> Version numbers should be used to avoid dependency hell... There's no reason for keeping the version number low, expect for marketing or maybe emotional attachment to versions. There's this concept in our mind that the version number has something to say about the project or the process or the stability and all of this is wrong... People believe something incredible existing is happening when a major/breaking version is released... Leave your emotions out of this.

He also has this good one in German: Hauptversionsnummernerhöhungsangst (fear of increasing major version).

Happy to see version 1.0 is coming :)

**Comment by NicolasRitouet** [comment]
I just released an alpha version for 1.0.0.
It is available under the `next` branch and is tagged `next` on npm (`npm install deployd@next`)
The changelog:
- removal of cli
- removal of dashboard
- removal of client-lib (the `dpd.js` file available in the frontend)

With this, the installation will be done through the `dpd-cli`:
```bash
npm i dpd-cli -g
dpd create hello
cd hello
dpd
```

We created 3 repos for these new modules:
- https://github.com/deployd/deployd-cli
- https://github.com/deployd/dpd-clientlib
- https://github.com/deployd/dpd-dashboard

What needs to be done before a final release:
- add unit / integration tests for the new modules
- test with various dpd modules
- more manual tests

Feel free to test and report/fix bugs.

**Comment by niallobrien** [comment]
Silly question as a new user: does Deployd generate the app code & rerun the server, or is everything maintained in the db and pulled from there? Eg. Routes etc. Thanks.

**Comment by andreialecu** [comment]
Only the data is in the database. All code is local to the project under the `resources` directory of the project.

**Comment by coderofsalvation** [comment]
my few cents:

* focus on progressive enhancement (add plugins for more features, but keep core simple)
* +1 for hjanuschka's focus & feature requests
* focus deploy is easy setting up api engine, datamanager included
* re-usable code, by introducing a new way of handling requests. Code needs to be specified as type 'simple' or 'middleware'

*simple:*
      this.foo = "bar"
      // current behaviour: when code ends, response is sent to client, runs in sandbox, editable in frontend. Not very re-usable / testable, but very convenient to setup a simple api in no time.

*middleware:*
      var dpd = this.dpd
      function(req,res,next){
        res.end({foo:"bar"})
      }
      // async / middleware style, this code is simply required, not sandboxed, and not editable in frontend. Very re-usable & testable for backend-ish purposes imho. The current hack for async request-code works..but looks..well..quite hacky :) This type of code also allows re-using code from the (big) express eco-system. Therefore plugins are easy to write.

Actually, if I wouldn't have existing deployservices in production, I would replace everything with middleware/async code. However, for backwards-compatible reasons, i would allow the developer to specify which code he wants to run/write.

* NOEDIT env-var: disable code-editing & adding/deletion of resources (on productionserver)
* support for middleware plugins (scan 'middleware'-array with modulenames in package.json )
      
      node_modules/dpd-middleware-foo/index.js
      ===============================

      dpd.use( function(req,res,next){  // alternatively: require('express-foo-middleware')
        next()
      },true)          // true means, insert middleware *after* collections-middleware

> REASON: this makes deployd highly patchable & modular & readable (the one-off modules only run in deployd context, which can be quite dauntning for newbies)

**Comment by rgolea** [comment]
@coderofsalvation Could you give examples for that last part? I might have not catched that last part properly.

**Comment by coderofsalvation** [comment]
sure!

Take for example this api-token middleware: https://github.com/coderofsalvation/dpd-api-token/blob/master/index.js#L50

The basic idea is: being able to cancel a request before it continues to the collection-middleware. 
Right now that's hard to do, since the router is hardwired to execute the collection-middleware first (hence the monkeypatch at line 50).
I took another approach for [https://github.com/coderofsalvation/dpd-acl-roles-permissions](dpd-acl-roles-permission) and decided to monkeypatch the mongodb-functions to be able to cancel them.

> NOTE: Sure you can integrate deployd into an express app, but i generally like the idea of drop-in modules for deployd (npm install dpd-modulename = all needed)

Maybe there are other ways of doing that, but i didn't find any other way to filter requests before they get processed by the default deployd-workflow.

Let me know if this makes any sense. Hope all is well
